Changes the build system to CMake

ilumsden opened the pull request:
Since @hariharan-devarajan mentioned that we wanted to do this and since I had some free time on the flight to SC, I just went ahead and started implementing a CMake-based build system for DYAD.

As of now, this PR just implements a drop-in replacement for autotools. It does **not** restructure the code base, add any major changes to the code, etc.

In case anyone else wants to contribute to this, I put the branch directly on this repo (as opposed to a fork).

ilumsden committed:
Initial work on converting DYAD from autotools to CMake

ilumsden committed:
Updates compile_test.yaml to use CMake instead of Autotools

ilumsden committed:
Typo fix

ilumsden committed:
Removes "VERSION" from find_package since it's not actually correct

hariharan-devarajan commented:
@ilumsden I will be looking into this actively. I will look at your branch and continue this work from here.

hariharan-devarajan committed:
refactored and initial changes to build.

hariharan-devarajan committed:
fixes for installation.

hariharan-devarajan committed:
fixed streaming apis.

hariharan-devarajan committed:
fixed installation script.

hariharan-devarajan committed:
fixed CI

hariharan-devarajan committed:
fixed CI

hariharan-devarajan committed:
fix ucx compile

hariharan-devarajan committed:
fix ci

hariharan-devarajan commented:
Did the following changes.

- [ ] Enabled CMake based Build
- [ ] Disabled Python test as its hanging.

hariharan-devarajan commented on CMakeLists.txt:
```diff
@@ -0,0 +1,231 @@
+# Build Type based configurations
+#------------------------------------------------------------------------------
+set(CMAKE_C_FLAGS ${CMAKE_C_FLAGS} -std=c99)
+set(CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS} -std=c++11)
```
Use minimum version 11 not absolute

hariharan-devarajan replied:
FIXED

hariharan-devarajan commented on CMake/dyad-config.cmake.build.in:
```diff
@@ -0,0 +1,18 @@
+#message(STATUS "DYAD_LIBRARY_PATH: " ${DYAD_LIBRARY_PATH})
+link_directories(${DYAD_LIBRARY_PATH})
+set(DYAD_LIBRARIES dlio-profiler)
+set(DYAD_DEFINITIONS "")
```
add alias dyad::dyad

hariharan-devarajan replied:
FIXED

hariharan-devarajan commented on CMake/dyad-config.cmake.build.in:
```diff
@@ -0,0 +1,18 @@
+set(DYAD_LIBRARIES dlio-profiler)
+set(DYAD_DEFINITIONS "")
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(dlio-profiler
```
change to dyad

hariharan-devarajan commented on CMake/dyad-config.cmake.install.in:
```diff
@@ -0,0 +1,17 @@
+#message(STATUS "DYAD_ROOT_DIR: " ${DYAD_ROOT_DIR})
+set(DYAD_LIBRARY_PATH "@DYAD_INSTALL_LIB_DIR@")
+link_directories(${DYAD_LIBRARY_PATH})
+set(DYAD_LIBRARIES dlio-profiler)
```
change to dyad

hariharan-devarajan replied:
FIXED

hariharan-devarajan commented on src/dyad/dtl/dyad_dtl_impl.c:
```diff
@@ -64,13 +65,14 @@ dyad_rc_t dyad_dtl_finalize (dyad_dtl_t **dtl_handle)
 #else
     if ((*dtl_handle)->mode == DYAD_DTL_FLUX_RPC) {
-#endif
+
```
Check here

hariharan-devarajan commented on src/dyad/wrapper/flux_barrier.c:
```diff
@@ -92,12 +93,12 @@ int main (int argc, char** argv)
-    strftime (tbuf, sizeof (tbuf), "%D %T", gmtime (&(t_now.tv_sec)));
-
-    FLUX_LOG_INFO ("Synchronized at %s.%09ld\n", tbuf, t_now.tv_nsec);
+//    struct timespec t_now;
```
Check here. This was not compiling on GCC 11

hariharan-devarajan committed:
separate setup of spack

hariharan-devarajan committed:
fix comments

hariharan-devarajan commented on CMake/dyad-config.cmake.build.in:
```diff
@@ -0,0 +1,18 @@
+set(DYAD_LIBRARY_PATH "@CMAKE_LIBRARY_OUTPUT_DIRECTORY@")
+#message(STATUS "DYAD_LIBRARY_PATH: " ${DYAD_LIBRARY_PATH})
+link_directories(${DYAD_LIBRARY_PATH})
+set(DYAD_LIBRARIES dlio-profiler)
```
FIXED

hariharan-devarajan committed:
fixed spack script

hariharan-devarajan committed:
fixed python

hariharan-devarajan committed:
Update compile_test.yaml

hariharan-devarajan committed:
Update compile_test.yaml

hariharan-devarajan committed:
Update compile_test.yaml

hariharan-devarajan commented:
@JaeseungYeom I fixed all the changes we talked about. Let me know if u think i should change anything else.

ilumsden commented on src/dyad/core/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+set(DYAD_CORE_SRC ${CMAKE_CURRENT_SOURCE_DIR}/dyad_core.c)
+set(DYAD_CORE_PRIVATE_HEADER ${CMAKE_CURRENT_SOURCE_DIR}/dyad_envs.h)
+set(DYAD_CORE_PUBLIC_HEADER ${CMAKE_CURRENT_SOURCE_DIR}/dyad_core.h)
```
If we don't want users directly calling DYAD Core, we should not have `dyad_core.h` in public headers. Alternatively, if we do want to install `dyad_core.h`, we need to add `dyad_dtl.h` to the public headers in the `dtl` directory.

hariharan-devarajan replied:
Fixed

ilumsden commented on src/dyad/dtl/CMakeLists.txt:
```diff
@@ -0,0 +1,43 @@
+set(DTL_SRC ${CMAKE_CURRENT_SOURCE_DIR}/dyad_dtl_impl.c)
+set(DTL_PRIVATE_HEADERS ${CMAKE_CURRENT_SOURCE_DIR}/dyad_dtl_impl.h
+                        ${CMAKE_CURRENT_SOURCE_DIR}/dyad_rc.h)
+set(DTL_PUBLIC_HEADERS )
```
See the comment for `CMakeLists.txt` in `core`

hariharan-devarajan replied:
We dont need it as core is not in public header

ilumsden commented on src/dyad/core/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+
+add_library(dyad_core SHARED ${DYAD_CORE_SRC})
+target_link_libraries(dyad_core PUBLIC Jansson::Jansson flux::core)
+target_link_libraries(dyad_core PRIVATE utils murmur3 dyad_dtl)
```
In the `dtl` directory, the CMake target is named `${PROJECT_NAME}_dtl`. We should use the same here

hariharan-devarajan replied:
Fixed

ilumsden commented on src/dyad/modules/CMakeLists.txt:
```diff
@@ -0,0 +1,42 @@
+add_library(dyad SHARED ${DYAD_MODULE_SRC})
+set_target_properties(dyad PROPERTIES PREFIX "")
+target_link_libraries(dyad PUBLIC Jansson::Jansson flux::core)
+target_link_libraries(dyad PRIVATE dyad_dtl)
```
In the `dtl` directory, the CMake target is named `${PROJECT_NAME}_dtl`. We should use the same here

hariharan-devarajan replied:
Fixed

hariharan-devarajan committed:
Changes to public header and libnames

1. Removed dyad_code from public headers
2. used PROJECT_NAME variable for all code
3. prefixed PROJECT_NAME for util code to avoid conflicts

JaeseungYeom commented:
Usually when I forgot to specify `CMAKE_INSTALL_PREFIX` with cmake command, I still make it work by editing the variable in `cmake_install.cmake` that is generated under build directory. However, I see `INSTALL DESTINATION` is fixed rather than depending on `CMAKE_INSTALL_PREFIX`. Can this behavior changed?

JaeseungYeom commented:
Also, can you confirm the python test is working?

hariharan-devarajan committed:
remove variables which prevent cmakr_install.cmake to use CMAKE_INSTALL_PREFIX

hariharan-devarajan commented:
> Usually when I forgot to specify `CMAKE_INSTALL_PREFIX` with cmake command, I still make it work by editing the variable in `cmake_install.cmake` that is generated under build directory. However, I see `INSTALL DESTINATION` is fixed rather than depending on `CMAKE_INSTALL_PREFIX`. Can this behavior changed?

I fixed it can u check it?

hariharan-devarajan committed:
Merge branch 'main' of github.com:flux-framework/dyad into cmake_refactor

 Conflicts:
	.github/workflows/compile_test.yaml

hariharan-devarajan commented:
> Also, can you confirm the python test is working?

It is working locally but not on CI.

JaeseungYeom commented:
> > Usually when I forgot to specify `CMAKE_INSTALL_PREFIX` with cmake command, I still make it work by editing the variable in `cmake_install.cmake` that is generated under build directory. However, I see `INSTALL DESTINATION` is fixed rather than depending on `CMAKE_INSTALL_PREFIX`. Can this behavior changed?
> 
> I fixed it can u check it?

Yes. It works.

JaeseungYeom commented:
Should python binding be installed as well?

JaeseungYeom commented on CMakeLists.txt:
```diff
@@ -0,0 +1,214 @@
+set_property(CACHE DYAD_CONTROL_PLANE PROPERTY STRINGS FLUX_RPC)
+set(DYAD_DATA_PLANE "FLUX_RPC" CACHE STRING "Protocol to use for DYAD's Data Plane")
+set_property(CACHE DYAD_DATA_PLANE PROPERTY STRINGS FLUX_RPC UCX)
+set(DYAD_PROFILER "NONE" CACHE STRING "Protocol to use for DYAD's Data Plane")
```
description of the option is irrelevant.

hariharan-devarajan replied:
Fixed

JaeseungYeom commented on CMakeLists.txt:
```diff
@@ -0,0 +1,214 @@
+#------------------------------------------------------------------------------
+# Build Type based configurations
+#------------------------------------------------------------------------------
+set(CMAKE_C_FLAGS ${CMAKE_C_FLAGS} -std=c99)
```
We should just use the consistent standard with c++

hariharan-devarajan replied:
when i switch the code to c++ we can move that right?

hariharan-devarajan replied:
FIXED

hariharan-devarajan committed:
fixed typo

hariharan-devarajan committed:
changed c99 to c11
